src/coordination/coordinator_instance.cpp

```diff
@@ -352,24 +315,57 @@ auto CoordinatorInstance::ShowInstances() const -> std::vector<InstanceStatus> {
 }
 
-auto CoordinatorInstance::ForceResetCluster_() -> ForceResetClusterStateStatus {
-  // Force reset tries to return cluster to state in which we have all the replicas we had before
-  // and try to do failover to new MAIN. Only then is force reset successful
+auto CoordinatorInstance::VerifyOrCorrectClusterState() -> VerifyOrCorrectClusterStateStatus {
+  utils::ExponentialBackoff backoff{std::chrono::milliseconds(1000), std::chrono::seconds(60)};
 
-  // 0. Open lock
-  // 1. Try to demote each instance to replica
-  // 2. Instances which are demoted proceed in next step as part of selection process
+  while (raft_state_->IsLeader()) {
+    spdlog::trace("Ensuring healthy state in cluster still as coordinator is leader and lock is opened.");
+    auto const result = VerifyOrCorrectClusterState_();
+    switch (result) {
+      case (VerifyOrCorrectClusterStateStatus::SUCCESS):
+        is_leader_ready_.store(true, std::memory_order_release);
+        spdlog::trace("Force reset was success.");
+        return result;
+      case VerifyOrCorrectClusterStateStatus::FAIL:
+        break;
+      case VerifyOrCorrectClusterStateStatus::SHUTTING_DOWN:
+        spdlog::trace("Stopping force as coordinator is shutting down.");
+        return result;
+    }
+    // TODO add stop token
+    backoff.wait();
+  }
+
+  return VerifyOrCorrectClusterStateStatus::SUCCESS;  // Shouldn't execute
+}
+
+auto CoordinatorInstance::VerifyOrCorrectClusterState_() -> VerifyOrCorrectClusterStateStatus {
+  // Only function which can run without opening lock as long as we have is_leader_ready_ set on false
+  // and all other functions checking whether leader is ready to start executing functions
+
+  // 0. Close lock
+  // 1. Get all instances from raft state, check if we have one main and others are replicas
+  // a) Set callbacks
+  // b) Start frequent checks
+  // c) Return true
+
+  // 2. Try find instance and do failover
   // 3. For selected instances try to send SWAP UUID and update log -> both must succeed
   // 4. Do failover
-  // 5. For instances which were down set correct callback as before
-  // 6. After instance gets back up, do steps needed to recover
+  // 5. For instances which are down set demote instance to main callback
 
-  spdlog::trace("Force resetting cluster in thread {}!", std::this_thread::get_id());
+  spdlog::trace("Doing VerifyOrCorrectClusterState.");
 
-  if (is_shutting_down_.load(std::memory_order_acquire)) {
-    return ForceResetClusterStateStatus::SHUTTING_DOWN;
+  if (is_shutting_down_.load()) {
+    return VerifyOrCorrectClusterStateStatus::SHUTTING_DOWN;
   }
 
+  // Do we need to open lock?
+  //  We don't need to open lock as long as we have check whether leader is ready on any action.
+  //  Also, function which uses this function can call in loop this function either until we succeed or we are leader.
+  //  If we have only three states (SUCCESS, FAIL, SHUTTING_DOWN) we can figure out whether we should continue executing
```

Comment:
Is it safe enough to rely only on is_leader_ready, you sure we don't need opened lock?

Reply:
It is fine, as mentioned on call